### PR TITLE
19 ivp c2

### LINKED ISSUE #19: Create schema for Initial Visit (IVP) form C2: Neuropsychological Battery Scores
https://www.alz.washington.edu/NONMEMBER/UDS/DOCS/VER3/IVPforms/c2_ivp.pdf

### PR DESCRIPTION
**Pull request by SenseCode**
close #19 
image
image
image

image

### CODE REVIEW

**Comment by mingjiecn** [comment]
Two designs for question options are used in redcap when enum is defined. Radio button is used when there are only a few options. Drop down is used when there are a long list for options. Some how when drop down design is used in redcap, there is no numbering in the front of the option. Without numbering in the front of the option, dictionary will have problem to find the matching option text. 
We have encountered this issue before. The way to fix it previously is to add numbering for those drop down design in redcap and in our schema.  Adding numbering to dropdown option in redcap is optional as long as rule for matching in redcap is provided to us. Do we need to talk about this at Tuesday meeting? @vsmalladi

**Comment by mingjiecn** [comment]
11/10/2020: The numbering rule can be find in Project Home and Design/Designer

**Comment by vsmalladi** [comment]
Possible to change pattern

            "pattern": "^(\\d+(\\.[1-9])?(\\-\\d+(\\.[1-9])?)?)|(unknown)$"

**Review by mingjiecn — COMMENTED** [review]
some patterns are not working. For example pattern ^[0-9]|[1-2][0-9]|30|88$ is not working when I check it with online regex tester. 

Also just beware that if you don't use enum for some options that has text content in it, then  those options will jut be shown with number no text. For example "88 item not administered will be just 88.

**Comment by SenseCode** [comment]
Take Venkat's idea, change Redcap dropdown list to text fields with pattern, super convenient. For future mapping issues  with numbers mixed with text options, I will suggest this way. @vsmalladi 
image
image
@mingjiecn Most of the fields have the 88 represent meaning in description. And I checked the original patient data, only numbers there for 88|95-98. I can edit more details if mapping have problems or we want to change. Feel free to let me know. Thanks for checking.

File: src/encoded/schemas/ivp_c2.json [hunk 1]
```diff
@@ -0,0 +1,1150 @@
...
+                    "enum": [
+                        "0 No",
+                        "1 Yes"
+                    ]
+                },
+                "mocatots": {
+                    "title": "1f. TOTAL RAW SCORE - UNCORRECTED (Not corrected for education or visual/hearing impairment)&nbsp ",
+                    "description": "((0-30, 88))(Enter 88 if any of the following MoCA items were not administered:  1g-1l, 1n-1t, 1w-1bb)",
+                    "type": "string",
+                  "pattern":"^[1-2]?[0-9]$|^30$|^88$"
```
Comment (mingjiecn) [inline src/encoded/schemas/ivp_c2.json:117]: Using anyof to include enum for key 88(88 item not administered)?

File: src/encoded/schemas/ivp_c2.json [hunk 2]
```diff
@@ -0,0 +1,1150 @@
...
+                },
+                "alert_mocatots": {
+                    "title": "WARNING - a value of 88 should be entered in the box above.  At least one of the following MoCA items was not administered: 1g-1l, 1n-1t, 1w-1bb.",
+                    "type": "string"
+                },
+                "mocatrai": {
+                    "title": "1g. Visuospatial/executive -- Trails",
+                    "description": "((0-1, 95-98))",
+                    "type": "string",
+                    "pattern":"^[0-1]$|^95$|^96$|^97$|^98$"
```
Comment (mingjiecn) [inline src/encoded/schemas/ivp_c2.json:127]: use anyof to include pattern for 1 -2 and enum for :
95, 95 = Physical problem
96, 96 = Cognitive/behavior problem
97, 97 = Other problem
98, 98 = Verbal refusal

So we won't lose readability.

File: src/encoded/schemas/ivp_c2.json [hunk 3]
```diff
@@ -0,0 +1,1150 @@
...
+                },
+                "mocacube": {
+                    "title": "1h. Visuospatial/executive -- Cube",
+                    "description": "((0-1, 95-98))",
+                    "type": "string",
+                    
+                    "pattern":"^[0-1]$|^95$|^96$|^97$|^98$"
```
Comment (mingjiecn) [inline src/encoded/schemas/ivp_c2.json:134]: use anyof to include pattern and enum

File: src/encoded/schemas/ivp_c2.json [hunk 4]
```diff
@@ -0,0 +1,1150 @@
...
+
+                },
+                "mocacloc": {
+                    "title": "1i. Visuospatial/executive -- Clock contour",
+                    "description": "((0-1, 95-98))",
+                    "type": "string",
+                    
+                    "pattern":"^[0-1]$|^95$|^96$|^97$|^98$"
```
Comment (mingjiecn) [inline src/encoded/schemas/ivp_c2.json:142]: use anyof to include pattern and enum

File: src/encoded/schemas/ivp_c2.json [hunk 5]
```diff
@@ -0,0 +1,1150 @@
...
+
+                },
+                "mocaclon": {
+                    "title": "1j.  Visuospatial/executive -- Clock numbers",
+                    "description": "((0-1, 95-98))",
+                    "type": "string",
+                    "pattern":"^[0-1]$|^95$|^96$|^97$|^98$"
```
Comment (mingjiecn) [inline src/encoded/schemas/ivp_c2.json:149]: use anyof to include pattern and enum

File: src/encoded/schemas/ivp_c2.json [hunk 6]
```diff
@@ -0,0 +1,1150 @@
...
+
+                },
+                "mocacloh": {
+                    "title": "1k. Visuospatial/executive -- Clock hands",
+                    "description": "((0-1, 95-98))",
+                    "type": "string",
+                    "pattern":"^[0-1]$|^95$|^96$|^97$|^98$"
```
Comment (mingjiecn) [inline src/encoded/schemas/ivp_c2.json:156]: use anyof to include pattern and enum

File: src/encoded/schemas/ivp_c2.json [hunk 7]
```diff
@@ -0,0 +1,1150 @@
...
+
+                },
+                "mocanami": {
+                    "title": "1l.  Language -- Naming",
+                    "type": "string",
+                    "pattern":"^[0-3]$|^95$|^96$|^97$|^98$"
```
Comment (mingjiecn) [inline src/encoded/schemas/ivp_c2.json:162]: use anyof to include pattern and enum

File: src/encoded/schemas/ivp_c2.json [hunk 8]
```diff
@@ -0,0 +1,1150 @@
...
+
+                },
+                "mocaregi": {
+                    "title": "1m. Memory -- Registration (two trials)",
+                    "description": "((0-10, 95-98))",
+                    "type": "string",
+                    "pattern":"^[0-9]$|^10$|^95$|^96$|^97$|^98$"
```
Comment (mingjiecn) [inline src/encoded/schemas/ivp_c2.json:169]: use anyof to include pattern and enum

File: src/encoded/schemas/ivp_c2.json (hunk 2 above)
Comment (SenseCode) [reply src/encoded/schemas/ivp_c2.json:127]: 1. This is a text box input number only, the meanings are already showed in the form instruction: see below.
image
image

2. I can add each property description with these interpretation, if we want.  

3.  If I change to  "pattern":"^[0-1]$|^95 = Physical problem$|^96 = Cognitive\/behavior problem$|^97 = Other problem$|^98 = Verbal refusal$", will the mapping still work?

**Comment by vsmalladi** [comment]
@SenseCode Some issues to resolve. Take a look at this once we merge everything else.

**Review by vsmalladi — APPROVED** [review]

**Pull request merged**